js/plugins/firebase/src/beta/data-connect.ts

```diff
@@ -0,0 +1,203 @@
+ * your connector's queries and mutations. This plugin is driven by the generated JSON file
+ * from the `llmTools` option. You can generate this file using the `llmTools` option in
+ * `connector.yaml`, for example:
+ *
+ * ```yaml
+ * connectorId: tools
+ * generate:
+ *   llmTools:
+ *     outputFile: ../../tools.json
+ * ```
+ *
+ * Once you have the tools file, you can use this function to register the tools as a Genkit
+ * plugin:
+ *
+ * ```ts
+ * const app = initializeFirebase({...});
+ *
+ * const ai = genkit({
+ *   plugins: [..., dataConnectTool({
+ *     name: 'myTools',
+ *     configFile: 'tools.json',
+ *     sdk: app,
+ *   })]
+ * })
+ * ```
+ *
+ * **IMPORTANT:** This plugin relies on the *client SDKs* for Firebase and does not have
+ * administrative access to Data Connect. To authenticate this plugin requires a `firebaseApp`
+ * instance on `context` with appropriate privileges or the use of the `firebaseContext()`
+ * context provider.
+ *
+ * @param options Configuration options for the plugin.
+ * @returns A Genkit plugin.
+ */
+export function dataConnectTools(
+  options: DataConnectToolsOptions
+): GenkitPlugin {
+  if (!options.config && !options.configFile)
+    throw new GenkitError({
+      status: 'INVALID_ARGUMENT',
+      message:
+        'Must supply `config` or `configFile` when initializing a Data Connect tools plugin.',
+    });
+
+  if (!options.config) {
+    try {
+      options.config = JSON.parse(
+        readFileSync(options.configFile, 'utf8')
+      ) as ToolsConfig;
+    } catch (e) {
+      throw new GenkitError({
+        status: 'INVALID_ARGUMENT',
+        message: `Could not parse Data Connect tools config from ${options.configFile}: ${(e as any).message}`,
+      });
+    }
+  }
+
+  return genkitPlugin(options.name, (ai: Genkit) => {
+    const config = options.config!;
+    for (const tool of config.tools) {
+      ai.defineTool(
+        {
+          name: `${options.name}/${tool.name}`,
+          description: tool.description || '',
+          inputJsonSchema: tool.parameters,
+        },
+        async (input, context) => {
+          const serverApp = serverAppFromContext(context, options.sdk);
+          const dc = getDataConnect(serverApp, {
+            connector: config.connector,
+            location: config.location,
+            service: config.service,
+          });
+
+          try {
+            if (tool.type === 'query') {
+              const { data } = await executeQuery(
+                queryRef(dc, tool.name, input)
+              );
+              return data;
+            }
+
+            const { data } = await executeMutation(
+              mutationRef(dc, tool.name, input)
+            );
+            logger.debug(
+              `[dataConnectTools] ${tool.name}(${JSON.stringify(input)}) -> ${JSON.stringify(data)}`
+            );
+            return data;
+          } catch (e) {
+            logger.info('[dataConnectTools] error on tool call:', e);
```

Comment:
```suggestion
            logger.error('[dataConnectTools] error on tool call:', e);
```

Reply:
Explicitly leaving this as an `info` because of the configurable behavior. FWIW I think we might want to adopt this `onError` for tools generally.